src/Aspire.Hosting/Extensions/ProjectResourceBuilderExtensions.cs

```diff
@@ -44,18 +43,136 @@ public static IResourceBuilder<ProjectResource> AddProject(this IDistributedAppl
+    /// <param name="projectPath">The path to the project file.</param>
+    /// <param name="launchProfileName">The launch profile to use. If <c>null</c> then no launch profile will be used.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<ProjectResource> AddProject(this IDistributedApplicationBuilder builder, string name, string projectPath, string? launchProfileName)
+    {
+        var project = new ProjectResource(name);
+
+        projectPath = PathNormalizer.NormalizePathForCurrentPlatform(Path.Combine(builder.AppHostDirectory, projectPath));
+
+        return builder.AddResource(project)
+                      .WithAnnotation(new ProjectMetadata(projectPath))
+                      .WithProjectDefaults(excludeLaunchProfile: launchProfileName is null, launchProfileName);
+    }
+
+    private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResourceBuilder<ProjectResource> builder, bool excludeLaunchProfile, string? launchProfileName)
     {
         // We only want to turn these on for .NET projects, ConfigureOtlpEnvironment works for any resource type that
         // implements IDistributedApplicationResourceWithEnvironment.
         builder.WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES", "true");
         builder.WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES", "true");
         builder.WithOtlpExporter();
         builder.ConfigureConsoleLogs();
+
+        if (excludeLaunchProfile)
+        {
+            builder.WithAnnotation(new ExcludeLaunchProfileAnnotation());
+            return builder;
+        }
+
+        if (!string.IsNullOrEmpty(launchProfileName))
+        {
+            builder.WithAnnotation(new LaunchProfileAnnotation(launchProfileName));
+        }
+
+        var projectResource = builder.Resource;
+
+        if (builder.ApplicationBuilder.ExecutionContext.IsRunMode)
+        {
+            // Automatically add EndpointAnnotation to project resources based on ApplicationUrl set in the launch profile.
+            var launchProfile = projectResource.GetEffectiveLaunchProfile(throwIfNotFound: true);
+            if (launchProfile is null)
+            {
+                return builder;
+            }
+
+            var urlsFromApplicationUrl = launchProfile.ApplicationUrl?.Split(';') ?? [];
+            foreach (var url in urlsFromApplicationUrl)
+            {
+                var uri = new Uri(url);
+
+                var endpointAnnotations = projectResource.Annotations.OfType<EndpointAnnotation>().Where(sb => string.Equals(sb.Name, uri.Scheme, StringComparisons.EndpointAnnotationName));
+                if (endpointAnnotations.Any(sb => sb.IsProxied))
+                {
+                    // If someone uses WithEndpoint in the dev host to register a endpoint with the name
```

Comment:
An end-developer cant get in this situation right? This could only happen if someone updates our code to call this method twice?

Reply:
Yeah, shouldn't be able to get here anymore. Can probably remove the code.